contrib/generate-ds20.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/python3
+# pylint: disable=invalid-name,missing-docstring
+#
+# Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+#
+# SPDX-License-Identifier: LGPL-2.1+
+#
+# pylint: disable=consider-using-f-string
+
+import sys
+import argparse
+import configparser
+import base64
+from typing import List
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-b",
+        "--bufsz",
+        type=int,
+        help="Buffer size in bytes",
+    )
+    parser.add_argument("-i", "--instance-id", type=str, help="Device instance ID")
+    parser.add_argument("filename", action="store", type=str, help="Quirk filename")
+
+    args = parser.parse_args()
+    if not args.bufsz:
+        parser.print_help()
+        sys.exit(1)
+
+    config = configparser.ConfigParser()
+    config.optionxform = str
+    try:
+        config.read(args.filename)
+    except configparser.MissingSectionHeaderError:
+        print("not a quirk file")
+        sys.exit(1)
+
+    # fall back to the default if there is only one device in the quirk file
+    if not args.instance_id:
+        sections = config.sections()
+        if len(sections) != 1:
+            print("multiple devices found, use --instance-id to choose")
+            sys.exit(1)
+        args.instance_id = sections[0]
+
+    # create the smallest kv store possible
+    lines: List[str] = []
+    try:
+        for key in config[args.instance_id]:
+            if key in ["Children", "Inhibit", "Issue", "Children"]:
+                print("WARNING: skipping key {}".format(key))
+                continue
+            value = config[args.instance_id][key]
+            lines.append("{}={}".format(key, value))
+    except KeyError:
+        print("no {} section".format(args.instance_id))
+        sys.exit(1)
+
+    # pad to the buffer size
+    buf: bytes = "\n".join(lines).encode()
+    if len(buf) > args.bufsz:
+        print("quirk data is larger than bufsz")
+        sys.exit(1)
+    buf = buf.ljust(args.bufsz, b"\0")
+
+    # success
+    print("DS20 descriptor control transfer data:")
+    print(" ".join(["{:02x}".format(val) for val in list(buf)]))
```

Comment:
Knowing a lot of people do their development on Windows and go back and forth what happens if you throw Windows line endings in here?  Does it get upset?  It might be worth stripping them first as a precaution.

Reply:
I think we need to strip on the python tool, and in the ds20 parser tbh

Reply:
The python tool didn't need anything, `configparser.ConfigParser()` just does the right thing. I've added a check to fail to load the descriptor if win32 line endings are detected; we want the descriptor to be as small as possible, and the parser to be as strict as possible. I've also added this as a note in the spec too. Something like `The UTF-8 quirk data must **not** contain Windows *CRLF-style* line endings.`

Reply:
Ah good.

> we want the descriptor to be as small as possible

So on this thought process, there is a few more possibilities that come to mind that can make it smaller:
1) Make the keys accept a shortened version as well.  IE instead of `Plugin` use `P`
2) Add a numerical assignment to all USB plugins and use that.  IE instead of `P=android_boot` use `P=2`
3) Use a compression algorithm for the buffer.  (With such small buffers this might not make a lot of sense.)
3) Use a numerical assignment for common patterns.  For example Logitech has identical `Flags`  for most devices.  So maybe there could be a number that represents that which gets encoded.

Any/all of the magic for that stuff I think could go into the `contrib` script.  So as a consumer of this you would write your quirk as normal, and the contrib script would do the best it could to shrink it.

Reply:
> Use a compression algorithm for the buffer

I actually tried zlib for this, but you're correct because the data is not repeated and because it's pretty tiny it actually increased the size by 34%!

I think we always want to support the unminimized quirk data, but I agree we can do something. I was thinking about something like fu_plugin_add_alias(plugin, "cap") where we could have a 3 letter acronym. Certainly flags could be specified by integer too, but that gets a bit messy as then the quirk "compiler" needs to know lots of grotty details about what the internal and plugin private flag values are. In most cases I think the control transfer is going to be of size 0x100 and that's a *lot* of data that can be included. I think we could also make flags "inherit" in the common case, so that setting one flag actually sets a bitmask of other flags -- e.g. `fu_device_alias_private_flag(FuDevice *self, guint64 value, const gchar *value_str)`

Reply:
I was talking to a vendor about this last week, and their consensus was it wasn't worth compressing and adding complexity in the docs. Basically their logic was they're already doing a 1024 byte transfer, and they don't care if it's `Plugin=tacom-usb` or `P=tac` as the other bytes will all be 0x0 and transferred anyway.

Reply:
OK, sounds good.  If we decide to change it later, it can always be a new minimum version requirement to use the minified version.

Reply:
Ack, agreed.